packages/redux-subspace-observable/src/observable/subspaced.js

```diff
@@ -16,15 +16,17 @@ const identity = (x) => x
 const subspaced = (mapState, namespace) => {
     const subspaceDecorator = subspace(mapState, namespace)
     return epic => (action$, store, dependencies) => {
-        const subspacedStore = subspaceDecorator(store)
+        const subspacedStore = subspaceDecorator(dependencies.store)
 
         const filteredAction$ = action$
             ::map((action) => subspacedStore.processAction(action, identity))
             ::filter(identity)
 
-        epic(filteredAction$, subspacedStore, dependencies).subscribe(subspacedStore.dispatch)
+        epic(filteredAction$, subspacedStore, { ...dependencies, store: subspacedStore })
```

Comment:
I wonder if we should be a bit more careful with the `store` param we pass on to the epic.  With this change the `subspaced` epic would still be able to dispatch into the store directly, bypassing the warning that was introduced.  Assuming this is still seen as an anti-pattern, I don't think we should swallow the warning.  This is more important for v1.0.0, where the dispatch function is expected to be missing, but a `subspaced` epic will still be able to dispatch happily.

Thoughts?

Reply:
Super great points!

As far as the second argument store `(action$, store, deps)` and not preventing the deprecation warning if they themselves call `store.dispatch`..I'm not sure how this will be possible. You may have to just duplicate similar deprecation logic lol.

Separately but related: I wouldn't pass on the subspacedStore in dependencies to users epics. The fact that you internally pass the store along dependencies should just be an implementation detail that doesn't impact users in any way. That would also mean that if users themselves pass a dependency with the key `store` it should still get passed along no matter what it is. I've helped some people do just that to temporarily future-proof their epics from the store's removal--its still discouraged, but at least then people are knowingly opting into the anti-pattern instead of us giving them the gun.

I think the solution to that is to store the users provided dependencies in a specific property instead of spreading it.

```js
baseCreateEpicMiddleware(rootEpic, { ...options, dependencies: { store, whatever: options.dependencies } })(store)
```

Then it can be extracted as-is and passed along to the users epics without knowing what type it is and not leaking the store in it.

Reply:
Whatever solution it won't have to live long. The second argument is changing to state$ as a stream in the coming weeks in v1.0.0-alpha

Reply:
@jayphelps 

> You may have to just duplicate similar deprecation logic lol.

Yeah, I'm thinking the same, but checking if `dispatch` is there first (i.e. if dispatch is on the store param, add a deprecation wrapper).  This was we will be compatible with current versions and future versions (we would probably remove it all together one day).

> I wouldn't pass on the subspacedStore in dependencies to users epics
> ...
> I think the solution to that is to store the users provided dependencies in a specific property instead of spreading it.

One of the features of redux subspace is being able to nest subspaces arbitrarily deep.  Because of this, we the nested subspace needs access to it's parent's subspace and use it the same way the parent subspace used the root store.

We also don't want to disrupt the `dependencies`that are passed too much as the rootEpic itself might have non-subspaced epics combined in.

For example:

```
const nestedEpic = (action$, store, dependencies) => { // uses dependencies.api }
const subspacedNestedEpic = subspaced((state) => state.nestedState, 'nested')(nestedEpic)

const standardMiddleEpic = (action$, store, dependencies) => { // uses dependencies.api }
const combinedMiddleEpic = combineEpics(standardMiddleEpic, subspacedNestedEpic)
const subspacedMiddleEpic = subspaced((state) => state.middleState, 'middle')(combinedMiddleEpic)

const standardRootEpic = (action$, store, dependencies) => { // uses dependencies.api }
const rootEpic = combineEpics(standardRootEpic, subspacedMiddleEpic)

const epicMiddleware = createEpicMiddleware(rootEpic, { dependencies: { api: { // create api } } })
const store = createStore(rootReducer, applyMiddleware(epicMiddleware))
```

Does that make sense?

> redux-observable won't pass undefined as the dependencies if they didn't pass the dependencies key in the options

I don't think this will be an issue for us as as far as redux-observable is concerned, there was always `dependencies` in the `options`.

> The second argument is changing to state$ as a stream in the coming weeks in v1.0.0-alpha

I saw this when looking for solutions to for this bug.  I'm not sure what this means for us.  I think we will need to map the `state$` stream and run the `mapState` function over it before passing it to the wrapped epic.  Does that sound feasible?

Reply:
> We also don't want to disrupt the dependencies that are passed too much as the rootEpic itself might have non-subspaced epics combined in.
> Does that make sense?

This one lost me actually. 😄  My suggestion was to not touch their dependencies which 

> I think we will need to map the state$ stream and run the mapState function over it before passing it to the wrapped epic. Does that sound feasible?

Not 100% sure what you mean but based on the code I've seen there shouldn't be a problem coming up with a solution. You'll have access to a stream of actions, a stream of state, and the latest state imperatively `state$.value` so I believe everything you would possibly need.

Reply:
> This one lost me actually. 😄 My suggestion was to not touch their dependencies which sounds the same as "don't want to disrupt the dependencies" to me lol

Yep, I can see why you're confused...  What I meant was we don't want to change the path the the existing dependencies as there might be a epic that has been combined that is not in a `subspaced` wrapper.

So in my (contrived) example, `standardRootEpic` expects to be able to access `api` directly on the dependencies object.  If we moved them into a `whatever` node as you suggested, this epic would need to know that and change to use `dependencies.whatever.api`.

Reply:
@mpeyper oh gotcha! hmm. I would place the store on a more private prop name then e.g. `__storeForSubspaceDontUseOrYouWillBeFired` to prevent it clashing with whatever they might provide for `store`. Chances are it would be the same store, but seems easy enough to avoid anyway.

Reply:
Prolly also noneumerable

```js
Object.defineProperty(dependencies, '__storeForSubspaceDontUseOrYouWillBeFired', {
  enumerable: false, // most important one to be a good citizen
  configurable: false,
  writable: false,
  value: store
});
```

Reply:
Anyways, sorry to jump in and start pushing my pedanticism 🕺

Reply:
So many things I did not know were even possible!

I think the `...DontUseOrYouWillBeFired` is a bit overkill.  Maybe something like `@@redux-subspace/store` so it doesn't even look like a key that could be used (I tested in jsfiddle and it seems to be allowed with `defineProperty`.

I think it will still need to writable so it can be replaced for nested subspaces, but other than that I like the idea.  Maybe we should do this in more places redux-subspace? 🤔

Reply:
@mpeyper lolol the prop name is totally your call, the name was just a joke based on the React one haha 😄 

https://github.com/facebook/react/blob/b2ca3349c27b57b1e9462944cbe4aaaf76783d2b
/src/React.js#L67

> React.__SECRET_DOM_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = ReactDOM;